scripts/build_osm_network.py

```diff
@@ -0,0 +1,1051 @@
+
+    return ac_freq_default
+
+
+def get_transformers(buses, lines):
+    """
+    Function to create fake transformer lines that connect buses of the same
+    station_id at different voltage.
+    """
+
+    ac_freq = get_ac_frequency(lines)
+    df_transformers = []
+
+    # Transformers should be added between AC buses only
+    # TODO pypsa-eur: Fix this! instead of tilde use !=
+    buses_ac = buses[buses["dc"] != True]
+    for g_name, g_value in buses_ac.sort_values("voltage", ascending=True).groupby(
+        by="station_id"
+    ):
+        # note: by construction there cannot be more that two buses with the same station_id and same voltage
+        n_voltages = len(g_value)
+
+        if n_voltages > 1:
+            for id in range(0, n_voltages - 1):
+                # when g_value has more than one node, it means that there are multiple voltages for the same bus
+                geom_trans = LineString(
+                    [g_value.geometry.iloc[id], g_value.geometry.iloc[id + 1]]
+                )
+
+                df_transformers.append(
+                    [
+                        f"transf_{g_name}_{id}",  # "line_id"
+                        g_value["bus_id"].iloc[id],  # "bus0"
+                        g_value["bus_id"].iloc[id + 1],  # "bus1"
+                        g_value.voltage.iloc[id],  # "voltage_bus0"
+                        g_value.voltage.iloc[id + 1],  # "voltage_bus0"
+                        g_value.country.iloc[id],  # "country"
+                        geom_trans,  # "geometry"
+                    ]
+                )
+    # TODO pypsa-eur: fix bug in pypsa-earth, where the id column is wrongly named "line_id" instead of "transformer_id
+    # name of the columns
+    trasf_columns = [
+        "transformer_id",
+        "bus0",
+        "bus1",
+        "voltage_bus0",
+        "voltage_bus1",
+        "country",
+        "geometry",
+    ]
+
+    df_transformers = gpd.GeoDataFrame(df_transformers, columns=trasf_columns)
+    if not df_transformers.empty:
+        init_index = 0 if lines.empty else lines.index[-1] + 1
+        df_transformers.set_index(init_index + df_transformers.index, inplace=True)
+    # update line endings
+    df_transformers = line_endings_to_bus_conversion(df_transformers)
+
+    return df_transformers
+
+
+def get_converters(buses):
+    """
+    Function to create fake converter lines that connect buses of the same
+    station_id of different polarities.
+    """
+
+    df_converters = []
+
+    for g_name, g_value in buses.sort_values("voltage", ascending=True).groupby(
+        by="station_id"
+    ):
+        # note: by construction there cannot be more that two buses with the same station_id and same voltage
+        n_voltages = len(g_value)
+
+        # A converter stations should have both AC and DC parts
+        if g_value["dc"].any() & ~g_value["dc"].all():
+            dc_voltage = g_value[g_value.dc]["voltage"].values
+
+            for u in dc_voltage:
```

Comment:
double-check if for-loop is performance bottleneck

Reply:
open todo